### ISSUE TITLE
Support dynamic addition and removal of virtual desktops as EWMH dictates

### ISSUE THREAD

**Issue by dumblob** [joewing/jwm#364]
The EWMH specification specifies two possible choices how an EWMH compliant window manager should react to a [_NET_NUMBER_OF_DESKTOPS](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472721968 ) message. Either reject it and do nothing else or honor it and behave as described in the linked section.

JWM currently rejects this request and this proposal is about honoring it instead.

**Steps to reproduce:**
1. Use the latest git revision of JWM or the release 2.3.7 and run it as a window manager.
1. Open terminal and run `wmctrl -n <number_of_your_virtual_desktops_increased_by_one>` or alternatively `wmctrl -n <number_of_your_virtual_desktops_decreased_by_one>`.

**What happens:**
Nothing.

**What was expected to happen:**
The number of virtual desktops should have increased (or decreased) by one (and thus also appear in the pager), which can be verified e.g. by `wmctrl -d`.

**Comment by technosaurus** [comment]
Since jwm cannot modify its own config file(s) or naturally communicate with its separate XML generator, this could be confusing and frustrating.  Currently you could add a binding that runs a program to do this via your XML generator, followed by a jwm -restart.

**Comment by Israel-** [comment]
I think refreshing individual items would be a feature that would be required before this could fully work.
Perhaps simply rereading the config and looking for only certain components specified (WindowStyle, Pager, Desktops, TrayButton, etc.....)
perhaps
`jwm -restart` could also take agruments optionally?
`jwm -restart Desktops`
Just a thought..

**Comment by dumblob** [comment]
@technosaurus 

>Since jwm cannot modify its own config file(s) or naturally communicate with its separate XML generator, this could be confusing and frustrating.

I'm afraid I don't quite understand what you mean. There is no need to change anything in XML nor in configuration file handling nor parsing nor absolutely anything else. The XML configuration JWM reads is an initial configuration. What happens afterwards (any dynamic behavior - e.g. windows are created/moved/removed, which is the primary motivation to even have a window manager) has nothing to do with the initial configuration any more - it lives it's own live.

>Currently you could add a binding that runs a program to do this via your XML generator, followed by a jwm -restart.

I'm confused. How do I react to the *_NET_NUMBER_OF_DESKTOPS* message by running a program (show me the code)?

@Israel- 

>Perhaps simply rereading the config and looking for only certain components specified (WindowStyle, Pager, Desktops, TrayButton, etc.....)

Many years ago (it could be already more than 11 :open_mouth:) I got a similar idea. But now I know it's a completely wrong path from the architectural point of view as it results in a lot of duplicating code or at least code doing similar things at more than one place, increases the size of the code base and is not easy to grasp.

Way easier is to understand the parsing, emitting and reacting to tokens/activities as a continually running machine, which waits all the time for instructions at the input, whereas at the very beginning we put the initial `jwmrc` file as the first set of instructions to be interpreted. Then the machine (the JWM window manager) shall wait (named pipe? socket? multiplatform_IPC_of_your_choice?) for new commands until the command `<Exit>` appears on the input. This is how GUIs are implemented, this is how many other window managers are implemented, etc.

**Comment by Israel-** [comment]
@dumblob 
I see, so a specific JWM system set in place to listen for things would be the way to go.  Some way of reading signals and reacting to them.  This would be a seemingly large undertaking, wouldn't it?  how do other WM/DE perform this, or what has been your thoughts on accomplishing this task in JWM?

This makes me think then that there should be a separate program to create dialogs (similar to gtk-dialog/yad/zenity/xmessage/etc...) that sends signals (pipe, socket, whatever....) to create a dialog.... or
I wonder if (in theory) JWM could run JWM in a window within itself and that with something like:
`jwm -f inception_dialog.xml`
IT would be really easy to build complicated interfaces using a jwmrc file, since buttons can run programs, and we can build a tray that acts as a menu.
Perhaps adding something like
```xml
<JWM windowed="true">
```
or

`jwm -window -f inception_dialog.xml`

JWM within JWM.

**Comment by dumblob** [comment]
> This would be a seemingly large undertaking, wouldn't it?

Well, not that large I would say as JWM seems already kind of decoupled. Of course considering only a small subset of all the instructions would be needed e.g. for the `_NET_NUMBER_OF_DESKTOPS` message while also assuming, that the X server will be used as a cheap and simple IPC.

What I'm rather afraid of is a potential porting effort to Wayland or so. Until JWM will use X server only for rendering and managing windows, it's still manageable to write a shim providing the X server like functionality using other tools (be it Wayland with few additional servers or whatever). But once JWM starts depending on the X server IPC, it could be harder to write the shim for a different IPC system.

>This makes me think then that there should be a separate program to create dialogs (similar to gtk-dialog/yad/zenity/xmessage/etc...) that sends signals (pipe, socket, whatever....) to create a dialog.... or
I wonder if (in theory) JWM could run JWM in a window within itself and that with something like:
jwm -f inception_dialog.xml
>...

This everything was already invented, programmed and used in the past - look e.g. at all the stuff developed for Plan 9 (especially the great window manager [Rio](https://en.wikipedia.org/wiki/Rio_(windowing_system) ) which was built around the idea of composition and nesting) or at Smalltalk or Lisp systems and other fully integrated systems, which allow basically any nesting and any change/configuration/extension with just few short lines of easily understandable code.

Btw. I'm not a fan of adding a "dialog-like" functionality to JWM. It's easier to add the IPC functionality and allow a "jclient" (JWM IPC client) to send nearly arbitrary commands and then use any of the existing dialog-like solutions to send these commands.

**Comment by inof** [comment]
Let me describe in detail how another favorite window manager of mine handles this. If jwm did that in a similar way, that would be really awesome.

Good old olvwm (open-look virtual window manager) supported that. I’m still using it on one of my machines, and I love that feature. There is a special window called VDM (virtual desktop manager). The manual page describes it as follows: _“[The VDM] is the window which represents visually the entire desktop: it provides a scaled-down version of the entire desktop. […] it will display a grid, each square of which maps to the size of the monitor. Each square is termed a logical screen.”_ So, the terminology is a bit confusing: What olvwm calls a “logical screen” is a “virtual desktop” (or just “desktop”) in jwm. What olvwm calls a “desktop” is the entirety of all logical screens (i.e. all virtual desktops connected together).

In other words, the VDM is somewhat similar to the pager in jwm, except that it’s placed in its own window and can be put anywhere on the screen, so it’s not restricted to a tray, and you can use it if you prefer not to have a tray at all (like I do).

And now here's the key point: When you resize the VDM window with the mouse (or with hotkeys), olvwm dynamically adds or removes virtual desktops accordingly. For example, suppose the size of your VDM is 4×1 squares (i.e. you have one row of 4 virtual desktops). Now you need more space, maybe just temporarily, to place additional terminal or browser windows or whatever. You just drag on a corner of the VDM and increase its width – causing olvwm to automatically add one more virtual desktop to the right, so now you’ve got 5×1. Similarly, you can increase the height to add one more row of squares, so now you’ve got 5×2.

Note that olvwm does not touch its configuration files. This is just a dynamic resizing at run-time. When the window manager is restarted, it re-reads its configuration and returns to the initial setting. If you want to change the number of virtual desktops permanently, you have to edit the configuration file yourself.

So, I see two steps for jwm:
* Support a visualization of all current desktops, placed in its own window (what olvwm calls “virtual desktop manager”, and ctwm calls this “workspace manager”). Currently, jwm has a featured called pager, but apparently it’s only available in the tray. Personally I prefer not to have a tray at all, so I cannot use the pager.
* Support dynamically adding / removing virtual desktops when the user resizes that window (let’s call it “pager window”) with the mouse.

**Comment by dumblob** [comment]
Indeed, two steps - first accept and react to the XDG standard messages with new/updated number of desktops. And second make it somehow integrated (e.g. the VDM functionality @inof described or the IPC client I proposed above).

@inof iIIRC n the past one could have multiple trays in JWM. Maybe this functionality (or bug?) is still in the lates version. In which case you can "fake" the VDM being "separate from a tray" by having only the pager in a separate tray with non-overlapping geometry.
